### PR TITLE
docs: add brief docs explaing udev's flock() block device node synchr…

### PR DESCRIPTION
**Pull request by poettering**
…onization

### CODE REVIEW

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+Note that `systemd-udevd` also watches all block device nodes it manages for
+`inotify()` `IN_CLOSE` events: whenever such an event is seen, this is used as
+trigger to re-run the rule-set for the device.
+
+These two concepts allow tools such as disk partitioners or file system
+formatting tools to safely and easily take exclusive ownership of a block
+device while operating: before starting work on the block device, they should
+take an `LOCK_EX` lock on it. This has two effects: first of all, in case
+`systemd-udevd` still processes the device the tool will wait for it to
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "still processes" → "is still processing"

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+finish. Secondly, after the lock is taken, it can be sure that that
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "Secondly" → "Second" (because "first" was used above).

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+`systemd-udevd` will refrain from processing the block device again, and thus
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: Why "again" here? It seems spurious.

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+all other client applications subscribed to it won't get device notifications
+from potentially half-written data either. After the operation is complete the
+partitioner/formatter can simply close the device node again. This has two
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: Here too, "again" seems unnecessary (it suggests that the tool is doing a second close, but it's not, most likely).

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+effects: it implicitly releases the lock, so that `systemd-udevd` can process
+events on the device node again. Secondly, it results in an `IN_CLOSE` event,
+which causes `systemd-udevd` to immediately re-process the device, seeing all
+changes the tool made, and notifying subscribed clients about it.
+
+Besides synchronizing block device access between `systemd-udevd` and such
+tools this scheme may also be used to synchronize access between those tools
+among each other, as the established Linux file system locking protocol is
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "among each other" → "themselves"

", as the established Linux file system locking protocol is simply reused for managing block device node access" → "" (I think it's more relevant that this is the same protocol that is used by udev, and the fact that this is a popular Linux protocol is not terribly important)

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+and might interfere with each other's work.
+
+Note that the file locks follow the usual access semantics of BSD locks: since
+`systemd-udevd` never writes to such block devices it only takes a `LOCK_SH`
+*shared* lock. A program intending to make changes to the block device should
+take a `LOCK_EX` *exclusive* lock instead. For further details, see the
+`flock(2)` man page.
+
+Summarizing: it is recommended to take `LOCK_EX` BSD file locks when
+manipulating block devices in all tools that manipulate file system block
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "that manipulate" → "that change" (to avoid repetition and clarify that this is about writing)

**Comment by poettering** [comment]
Force pushed a new version now. I addressed the points @keszybz raised. I also made one other change: there's a new paragraph highlighting that POSIX and BSD file locks are different concepts, and emphasizing we use the latter.

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+device while operating: before starting work on the block device, they should
+take an `LOCK_EX` lock on it. This has two effects: first of all, in case
+`systemd-udevd` is still processing the device the tool will wait for it to
+finish. Second, after the lock is taken, it can be sure that that
+`systemd-udevd` will refrain from processing the block device, and thus all
+other client applications subscribed to it won't get device notifications from
+potentially half-written data either. After the operation is complete the
+partitioner/formatter can simply close the device node. This has two effects:
+it implicitly releases the lock, so that `systemd-udevd` can process events on
+the device node again. Secondly, it results in an `IN_CLOSE` event, which
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "in an" → "an"

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+causes `systemd-udevd` to immediately re-process the device, seeing all changes
+the tool made, and notifying subscribed clients about it.
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "notifying" → "notify" (because it's for "causes systemd-udev to ...")

File: docs/BLOCK_DEVICE_LOCKING.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+Note that the file locks follow the usual access semantics of BSD locks: since
+`systemd-udevd` never writes to such block devices it only takes a `LOCK_SH`
+*shared* lock. A program intending to make changes to the block device should
+take a `LOCK_EX` *exclusive* lock instead. For further details, see the
+`flock(2)` man page.
+
+And please keep in mind: BSD file locks (`flock()`) and POSIX file locks
+(`lockf()`, `F_SETLK`, …) are different concepts, and in their effect
+orthogonal. The scheme discussed above uses the former and not the latter
+concept, because the these types of locks more closely match the required
```
Comment (keszybz) [inline docs/BLOCK_DEVICE_LOCKING.md]: "concept" → ""

File: docs/BLOCK_DEVICE_LOCKING.md (hunk 7 above)
Comment (poettering) [reply docs/BLOCK_DEVICE_LOCKING.md]: "… results an `IN_CLOSE` …"? Is that proper English?

**Comment by poettering** [comment]
force pushed a new version, with the three other comments addressed (though "results an" sounds wrong to me)

File: docs/BLOCK_DEVICE_LOCKING.md (hunk 7 above)
Comment (keszybz) [reply docs/BLOCK_DEVICE_LOCKING.md]: You're right, I got confused by too many "in"s here.

**Pull request merged**